includes/class-wc-database-locks.php

```diff
@@ -0,0 +1,131 @@
+<?php
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly
+}
+
+/**
+ * Allows the DB to be locked to avoid concurrent events doing the same thing.
+ *
+ * @since    3.2.0
+ * @package  WooCommerce/Classes
+ * @category Class
+ * @author   Automattic
+ */
+class WC_Database_Locks {
+
+	/**
+	 * Request ID.
+	 * @var string
+	 */
+	private static $request_id = '';
+
+	/**
+	 * Get ID of this request.
+	 *
+	 * @return string
+	 */
+	public static function get_request_id() {
+		if ( ! self::$request_id ) {
+			self::$request_id = uniqid( wp_rand(), true );
+			add_action( 'shutdown', __CLASS__, 'release_all_locks' );
```

Comment:
To prevent something from permanently locking, we would/should have something that runs and looks for expired locks and releases based on that. i.e., on `shutdown`: `release_expired_locks`. 

If we tie these to a specific request, a default `$release_timeout` of an hour seems really high.